stargz/stargzify/stargzify.go

```diff
@@ -25,11 +27,121 @@ import (
 	"github.com/google/go-containerregistry/pkg/v1/tarball"
 )
 
+var (
+	upgrade = flag.Bool("upgrade", false, "upgrade the image in-place by overwriting the tag")
+
+	usage = `Usage: %[1]s [-upgrade] input [output]
+
+Converting images:
+  # converts "ubuntu" from dockerhub and uploads to your GCR project
+  %[1]s ubuntu gcr.io/<your-project>/ubuntu:stargz
+
+  # converts and overwrites :latest
+  %[1]s -upgrade gcr.io/<your-project>/ubuntu:latest
+
+Converting files:
+  %[1]s file:/tmp/input.tar.gz file:output.stargz
+
+  # writes to /tmp/input.stargz
+  %[1]s file:/tmp/input.tar.gz
+`
+)
+
 func main() {
-	if len(os.Args) < 3 {
-		log.Fatalf("usage: %s original-image converted-image", os.Args[0])
+	flag.Parse()
+	if len(flag.Args()) < 1 {
+		printUsage()
+	}
+
+	if strings.HasPrefix(flag.Args()[0], "file:") {
+		// We'll use "file:" prefix as a signal to convert single files.
+		convertFile()
+	} else {
+		convertImage()
+	}
+}
+
+func printUsage() {
+	log.Fatalf(usage, os.Args[0])
+}
+
+func convertFile() {
+	var in, out string
+	if len(flag.Args()) > 0 {
+		in = strings.TrimPrefix(flag.Args()[0], "file:")
 	}
-	src, dst := os.Args[1], os.Args[2]
+	if len(flag.Args()) > 1 {
+		out = strings.TrimPrefix(flag.Args()[1], "file:")
+	}
+
+	var f, fo *os.File // file in, file out
+	var err error
+	switch in {
+	case "":
+		printUsage()
+	case "-":
+		f = os.Stdin
+	default:
+		f, err = os.Open(in)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	defer f.Close()
+
+	if out == "" {
+		if in == "-" {
+			out = "-"
+		} else {
+			base := strings.TrimSuffix(in, ".gz")
+			base = strings.TrimSuffix(base, ".tgz")
+			base = strings.TrimSuffix(base, ".tar")
+			out = base + ".stargz"
+		}
+	}
+	if out == "-" {
+		fo = os.Stdout
+	} else {
+		fo, err = os.Create(out)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	w := stargz.NewWriter(fo)
+	if err := w.AppendTar(f); err != nil {
+		log.Fatal(err)
+	}
+	if err := w.Close(); err != nil {
+		log.Fatal(err)
+	}
+	if err := fo.Close(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func parseFlags(args []string) (string, string) {
+	if len(args) < 1 {
+		printUsage()
+	}
+
+	var src, dst string
+	src = args[0]
+
+	if len(args) < 2 {
+		if *upgrade {
+			dst = src
+		} else {
+			printUsage()
+		}
+	} else {
+		dst = args[1]
```

Comment:
You should also give a usage message if --upgrade is used with len(args) == 2

and complain in general if len(args) > 2